File: contracts/admin/sessionspans.go
package admin

import "time"

// SessionSpan is one element of the SessionSpansDTO v1 array returned by
// GET /api/v1/sessions/{id}/spans — one gen_ai span (one upstream request) of
// a session, projected from the request_events span_event blob alone (never
// from Records; invariant #4). The shape is frozen by the span-dto JSON schema
// (sluice:session-spans-dto/v1) the session lifecycle page renders from.
//
// Truncation policy: this server serves FULL text and tool arguments — no
// cap is applied — and every *Chars field carries the same value as the
// served text's length, so the renderer's "showing first N of M" guard never
// fires. The *Chars fields remain in the contract because the schema makes
// truncation a server prerogative; a future cap only changes the server, not
// the renderer.
type SessionSpan struct {
	// CID is the correlation id (= span id in the console).
	CID string `json:"cid"`

	// At is the request receipt time (the span start), RFC3339.
	At time.Time `json:"at"`

	// LatencyMs is the request wall time; null when the span blob is absent.
	LatencyMs *int64 `json:"latency_ms"`

	// TTFCMs is gen_ai.response.time_to_first_chunk converted to ms; null for
	// non-streaming requests (the attribute is emitted on streams only).
	TTFCMs *int64 `json:"ttfc_ms"`

	// Status is the upstream HTTP status (sluice.upstream_status, falling back
	// to the client-facing status column); null when neither is known.
	Status *int `json:"status"`

	// Model is the requested model; null when unknown.
	Model *string `json:"model"`

	// FinishReason is the first gen_ai.response.finish_reasons entry; null
	// when the response carried none.
	FinishReason *string `json:"finish_reason"`

	// SessionID is the session bundle root every span of the session shares.
	SessionID string `json:"session_id"`

	// ConversationID is the thread the request belongs to: the session for the
	// main loop, the subagent thread id for sub-agents.
	ConversationID string `json:"conversation_id"`

	// ParentConversationID links a subagent thread toward its parent; null for
	// the main loop.
	ParentConversationID *string `json:"parent_conversation_id"`

	// Usage is the per-span token rollup incl. the server-tool counter map.
	Usage SessionSpanUsage `json:"usage"`

	// OutputParts is the normalized assistant output envelope, in emission
	// order. Always an array (empty when content capture was off).
	OutputParts []SessionSpanOutputPart `json:"output_parts"`

	// InputParts is the trailing input delta — the NEW parts of this request
	// only. Always an array (empty when content capture was off).
	InputParts []SessionSpanInputPart `json:"input_parts"`

	// InputText is the human text of the trailing message when InputParts has
	// text parts (concatenated in order); null otherwise.
	InputText *string `json:"input_text"`

	// InputTextChars is the uncapped total size of InputText.
	InputTextChars *int `json:"input_text_chars"`

	// OutputText is the concatenated output text parts; null when the
	// response carried no text part.
	OutputText *string `json:"output_text"`

	// OutputTextChars is the uncapped total size of OutputText.
	OutputTextChars *int `json:"output_text_chars"`
}

// SessionSpanUsage is the per-span token rollup. Each count is null when the
// span carried no such usage attribute (e.g. an errored request with no usage
// block), distinguishing "not reported" from a genuine zero.
type SessionSpanUsage struct {
	// Input / Output are gen_ai.usage.input_tokens / .output_tokens.
	Input  *int64 `json:"input"`
	Output *int64 `json:"output"`

	// CacheRead / CacheCreation are the provider prompt-cache counts
	// (gen_ai.usage.cache_read.input_tokens / .cache_creation.input_tokens).
	CacheRead     *int64 `json:"cache_read"`
	CacheCreation *int64 `json:"cache_creation"`

	// ServerToolUse is the generic provider-executed tool counter map
	// (gen_ai.usage.server_tool_use.*), keyed by the provider's own counter
	// name; null when the span carried none.
	ServerToolUse map[string]int64 `json:"server_tool_use"`
}

// SessionSpanOutputPart is one normalized assistant output part. Type is one
// of text | reasoning | tool_call | tool_call_response | unknown; only the
// fields relevant to the type are set.
type SessionSpanOutputPart struct {
	// Type is the part discriminator.
	Type string `json:"type"`

	// Chars is the true size of the part's text (text / reasoning content, or
	// a tool_call_response's result). Absent for tool_call parts.
	Chars *int `json:"chars,omitempty"`

	// ID is the tool_call / tool_call_response join key.
	ID string `json:"id,omitempty"`

	// Name is the tool name (tool_call only).
	Name string `json:"name,omitempty"`

	// Args is the raw tool-call arguments JSON — an opaque vendor payload the
	// renderer never field-picks. Served whole (no cap).
	Args string `json:"args,omitempty"`

	// ArgsChars is the uncapped arguments size; equal to len(Args) since this
	// server applies no cap.
	ArgsChars *int `json:"args_chars,omitempty"`

	// Text is the result text of a tool_call_response output part (a
	// server-executed tool's result, carried on the same span as its call).
	// Served whole (no cap); Chars carries its size.
	Text string `json:"text,omitempty"`
}

// SessionSpanInputPart is one part of the trailing input delta. Type is
// text | tool_call_response.
type SessionSpanInputPart struct {
	// Type is the part discriminator.
	Type string `json:"type"`

	// ID joins a tool_call_response back to a prior span's tool_call.
	ID string `json:"id,omitempty"`

	// Chars is the true total size of Text before any cap (equal to
	// len(Text) since this server applies no cap).
	Chars *int `json:"chars,omitempty"`

	// Text is the part's own text — the user message text or the tool result
	// body. Served whole (no cap).
	Text string `json:"text,omitempty"`

	// IsError marks an errored tool result. Omitted today: the gateway's
	// content normalizer does not yet capture the provider is_error flag.
	IsError bool `json:"is_error,omitempty"`
}

File: web/src/lib/session-spans.ts
// SessionSpansDTO client. Types mirror span-dto.schema.json (SessionSpansDTO
// v1) — the expected response of GET /api/v1/sessions/{id}/spans: the flat
// gen_ai span list the session lifecycle page renders from. Every view-model
// value (lanes, gaps, the tool ledger, exec stats) is derived from this shape
// at render time per the renderer contract; the page has no knowledge of
// Records or the archive.

import { apiFetch } from "@/lib/api"

// SpanUsage is the per-span token rollup. server_tool_use is the generic
// counter map (gen_ai.usage.server_tool_use.*) for provider-executed tools.
export type SpanUsage = {
  input?: number | null
  output?: number | null
  cache_read?: number | null
  cache_creation?: number | null
  server_tool_use?: Record<string, number> | null
}

// OutputPartType enumerates the normalized assistant-output part envelope.
export type OutputPartType = "text" | "reasoning" | "tool_call" | "tool_call_response" | "unknown"

// OutputPart is one normalized assistant output part, in emission order.
// `args` is the raw arguments JSON, server-capped; it is an opaque vendor
// payload the renderer must never field-pick. `args_chars` carries the
// uncapped size — greater than args.length means the server truncated.
export type OutputPart = {
  type: OutputPartType
  chars?: number
  id?: string
  name?: string
  args?: string
  args_chars?: number
}

// InputPart is one part of the trailing input delta (the NEW parts of this
// request only). A tool_call_response's `id` joins to a prior span's
// tool_call id; `chars` is the true total size before any server cap.
export type InputPart = {
  type: "text" | "tool_call_response"
  id?: string
  chars?: number
  text?: string
  is_error?: boolean
}

// SessionSpan is one gen_ai span of the session — one upstream request.
export type SessionSpan = {
  // cid is the correlation id (= span id in the console).
  cid: string
  // at is the request receipt time, RFC3339.
  at: string
  latency_ms?: number | null
  // ttfc_ms is gen_ai.response.time_to_first_chunk in ms (null when
  // non-streaming).
  ttfc_ms?: number | null
  // status is the upstream HTTP status.
  status?: number | null
  model?: string | null
  finish_reason?: string | null
  session_id: string
  // conversation_id = session_id for the main loop, AgentID for sub-agents.
  conversation_id: string
  parent_conversation_id?: string | null
  usage: SpanUsage
  output_parts: OutputPart[]
  input_parts: InputPart[]
  // input_text is the human text of the trailing message when input_parts has
  // a text part, server-capped; input_text_chars carries the uncapped total.
  input_text?: string | null
  input_text_chars?: number | null
  // output_text is the concatenated output text parts, server-capped;
  // output_text_chars carries the uncapped total.
  output_text?: string | null
  output_text_chars?: number | null
}

// SessionSpansResult tags the span list with where it came from: "api" is the
// live projection; "fixture" is the bundled synthetic fallback served when the
// backend has no /spans endpoint yet, so the page can label the data honestly.
export type SessionSpansResult = {
  spans: SessionSpan[]
  source: "api" | "fixture"
}

/**
 * Fetches the session's gen_ai span list for the lifecycle page. When the
 * endpoint 404s (a backend that predates the spans projection), falls back to
 * the bundled synthetic fixture — lazily imported so it stays out of the main
 * chunk — and says so via `source`. Anything else (incl. UnauthorizedError)
 * bubbles up.
 */
export async function fetchSessionSpans(id: string): Promise<SessionSpansResult> {
  try {
    const spans = await apiFetch<SessionSpan[]>(`/api/v1/sessions/${encodeURIComponent(id)}/spans`)
    return { spans: spans ?? [], source: "api" }
  } catch (err) {
    if ((err as { status?: number }).status === 404) {
      const { SESSION_SPANS_FIXTURE } = await import("@/telemetry/mock/session-spans-fixture")
      return { spans: SESSION_SPANS_FIXTURE, source: "fixture" }
    }
    throw err
  }
}
